### DISCUSSION TITLE
How to use `LoadImage`

### DISCUSSION THREAD

**Discussion by karmeye** [microsoft/CsWin32#1171] [Q&A] [answered]
I'm trying to load an icon to set in the Windows taskbar, using this code:

```cs
public static void LoadIcon(Window window, string iconName)
{
	string iconPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, iconName);

	if (!File.Exists(iconPath))
		return;

	var hwnd = new HWND(GetWindowHandle(window));

	using var hIcon = PInvoke.LoadImage(null, iconPath, GDI_IMAGE_TYPE.IMAGE_ICON, 128, 128, IMAGE_FLAGS.LR_LOADFROMFILE);

	var handle = hIcon.DangerousGetHandle();
	if (handle != 0)
	{
		PInvoke.SendMessage(hwnd, PInvoke.WM_SETICON, new WPARAM(PInvoke.ICON_BIG), new LPARAM(handle));
	}

	//PInvoke.DestroyIcon(handle);
}
```

- Whether I call `SendMessage` or not, I get `SEHException`.
- If I do not dispose (with using) I eventually get a exception with message: "External component has thrown an exception."
- Is it related to https://github.com/microsoft/CsWin32/issues/715 ?

Is there any workaround?

Thanks

**Comment by AArnott** [comment]
I wonder if you're expected to keep the icon valid for the life of it being displayed on the taskbar, rather than disposing of it. I also see you calling `DangerousGetHandle` without calling `DangerousAddRef`, which may be a problem for non-deterministic disposal by the runtime.
Your pointing out #715 is apt. You should probably avoid the `SafeHandle` overload altogether.

**Reply by karmeye** [reply]
Say I needed to keep it retain for the lifetime of the application, how would I correctly get the handle of the icon and then, how would I dispose it when the application exits?

**Reply by AArnott** [reply, marked as answer]
If you avoid the `SafeHandle` overload altogether, there won't be a `SafeHandle` to worry about releasing or finalizing. The native `HANDLE` you get back would typically need to be released when you're done with it, but if you want it for the lifetime of your application, you don't ever have to release it because closing the process has the OS implicitly release it. So you can call the `LoadImage` overload that returns a HANDLE and then just drop the result (after using it in SendMessage).
